package/MDAnalysis/core/universe.py

```diff
@@ -530,71 +541,140 @@ def Merge(*args):
+            if type(attr) != np.ndarray:
+                raise TypeError('Encountered unexpected topology'+
+                                'attribute of type {}'.format(
+                                type(attr)))
+            try:
+                attr_array.extend(attr)
+            except NameError:
+                attr_array = list(attr)
+        attrs.append(attr_class(np.array(attr_array,
+                                        dtype=attr.dtype)))
+        del attr_array
+
+    # Build up topology groups
+    for tg in (topology_groups & common_attrs):
+        bondidx = []
+        types = []
+        offset = 0
+        for ag in disjoint_atom_groups:
+            bonds = getattr(ag, tg)
+            bond_class = type(getattr(ag.universe._topology, tg))
+            bondidx.extend(bonds.indices + offset)
+            if hasattr(bonds, '_bondtypes'):
+                types.extend(bonds.types())
+            else:
+                types.extend([None]*len(bonds))
+            offset += len(ag)
+        bondidx = np.array(bondidx, dtype=np.int32)
+        if any(t is None for t in types):
+            attrs.append(bond_class(values))
+        else:
+            types = np.array(types, dtype='|S8')
+            attrs.append(bond_class(bondidx, types))
+
+    # Renumber residue and segment indices
+    n_atoms = sum([len(ag) for ag in disjoint_atom_groups])
+    residx = []
+    segidx = []
+    for ag in disjoint_atom_groups:
+        res_offset = len(set(residx))
+        resdict = {n: i+res_offset for i, n in enumerate(set(ag.resindices))}
+        seg_offset = len(set(segidx))
+        segdict = {n: i+len(set(segidx)) for i, n in enumerate(set(
+                                                     ag.segindices))}
+        residx.extend([resdict[n] for n in ag.resindices])
+        segidx.extend([segdict[n] for n in ag.segindices])
+
+    residx = np.array(residx, dtype=np.int32)
+    segidx = np.array(segidx, dtype=np.int32)
+
+    n_residues = len(set(residx))
+    n_segments = len(set(segidx))
+
+    top = Topology(n_atoms, n_residues, n_segments,
+                   attrs=attrs,
+                   atom_resindex=residx,
+                   residue_segindex=segidx)
+
+    # Create blank Universe and put topology in it
+    u = Universe()
+    u._topology = top
+
+    # generate Universe version of each class
+    # AG, RG, SG, A, R, S
+    u._classes = groups.make_classes()
+
+    # Put Group level stuff from topology into class
+    for attr in u._topology.attrs:
+        u._process_attr(attr)
+
+    # Generate atoms, residues and segments
+    u.atoms = u._classes['atomgroup'](
+            np.arange(u._topology.n_atoms), u)
+
+    u.residues = u._classes['residuegroup'](
+            np.arange( u._topology.n_residues), u)
+
+    u.segments = u._classes['segmentgroup'](np.arange(
+        u._topology.n_segments), u)
+
+    # Update Universe namespace with segids
+    for seg in u.segments:
```

Comment:
So with things like this, it's much better to move the original version into a method (something like `_make_segid_namespace`), and then call that method from the two places it gets called from.  Then if we change how this works it gets updated everywhere